### PR TITLE
Feature/pre substitute transforms

### LINKED ISSUE #5142: Perform Environment Variable Substitution Prior to Other Transforms
# Use Case
When using a tool like [mozilla/sops](https://github.com/mozilla/sops) to safely store your secrets in source control, each environment has its own secrets file.  In the base `app-config.yaml`, it would be convenient to be able to declare something like below, rather than having to have it duplicated in every single environment YAML file (i.e. `app-config.local.yaml`, `app-config.dev.yaml`, `app-config.prod.yaml`), with only the name of the file differing.

```yaml
integrations:
  github:
    - host: github.com
      apps:
        - $include: config/backstage.${BACKSTAGE_ENVIRONMENT}.secrets.yaml#githubApp

```

It would seem that adding a call to the substitution transformer at this point, and using the resulting value (if the transform returned `applied: true`) would work for this use case.
https://github.com/backstage/backstage/blob/95ac3cd52c36ec015d6a64936e5da3d4a13b8c17/packages/config-loader/src/lib/transform/include.ts#L56

This would require passing in the substitution transformer to the include transformer (or creating a new instance...the former seems like a better option).  If this seems like a reasonable approach, I don't mind tackling the change sometime in the next week or two.

### PR DESCRIPTION
**Pull request by JP-Dhabolt**
## Hey, I just made a Pull Request!

* Fixed bug where `$${...}` was not being escaped to `${...}` as per the docs
* Added support for using Environment Variable Substitution in the includes transforms, to support the below config:

```yaml
integrations:
  github:
    - host: github.com
      apps:
        - $include: secrets.${BACKSTAGE_ENVIRONMENT}.yaml
```

Resolves #5142.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/config-loader/src/lib/transform/substitution.ts [hunk 1]
```diff
@@ -29,8 +29,13 @@ export function createSubstitutionTransform(env: EnvFunc): TransformFunc {
...
-    for (let i = 1; i < parts.length; i += 2) {
-      parts[i] = await env(parts[i]!.trim());
+    for (let i = 0; i < parts.length; i++) {
+      if (i % 2 === 0) {
+        parts[i] = parts[i]!.replace(/\$\${(.*?)\}/, '${$1}');
+      }
+      if (i % 2 === 1) {
+        parts[i] = await env(parts[i]!.trim());
+      }
     }
```
Comment (Rugvip) [inline packages/config-loader/src/lib/transform/substitution.ts:39]:
```suggestion
    const parts: (string | undefined)[] = input.split(/(\$?\$\{[^{}]+\})/);
    for (let i = 1; i < parts.length; i += 2) {
      const part = parts[i]!;
      if (part.startsWith('$$')) {
        parts[i] = part.slice(1);
      } else {
        parts[i] = await env(part.slice(2, -1).trim());
      }
    }
```

Thinking it's nice to skip the extra replace

File: packages/config-loader/src/lib/transform/include.ts [hunk 2]
```diff
@@ -53,7 +54,11 @@ export function createIncludeTransform(
       return { applied: false };
     }
 
-    const includeValue = input[includeKey];
+    const rawIncludedValue = input[includeKey];
+    const substituteResults = await substitute(rawIncludedValue!, baseDir);
+    const includeValue = substituteResults.applied
+      ? substituteResults.value
+      : rawIncludedValue;
```
Comment (Rugvip) [inline packages/config-loader/src/lib/transform/include.ts:61]: Let's move this down after the string check below, and then check again whether we got `undefined` back from the substitution, and in that case throw a bit more descriptive error.

File: packages/config-loader/src/lib/transform/include.test.ts [hunk 3]
```diff
@@ -14,25 +14,47 @@
...
+const substituteMe = '${MY_SUBSTITUTION}';
+const mySubstitution = 'fooSubstitution';
 
 const env = jest.fn(async (name: string) => {
   return ({
     SECRET: 'my-secret',
   } as { [name: string]: string })[name];
 });
 
+const substitute = jest.fn(
```
Comment (Rugvip) [inline packages/config-loader/src/lib/transform/include.test.ts:32]: Could skip making this a mocked function since we're not asserting anything

**Review by Rugvip — COMMENTED** [review]
Makes sense, thank you! 👍 

Just a couple of nits and suggestions

File: packages/config-loader/src/lib/transform/substitution.ts (hunk 1 above)
Comment (JP-Dhabolt) [reply packages/config-loader/src/lib/transform/substitution.ts]: Makes sense.  I applied this, but had to change the `+` in the regex to `*`, otherwise the test for `$${}` being escaped to `${}` failed, as it has no characters in the brackets.

I added 2 additional tests for this as well:
* Test to make sure that `${}` returns `undefined`, since this will now match with the updated regex.
* Test to make sure that multiple escapes in a row are handled correctly.

File: packages/config-loader/src/lib/transform/include.ts (hunk 2 above)
Comment (JP-Dhabolt) [reply packages/config-loader/src/lib/transform/include.ts:65]: Applied.

File: packages/config-loader/src/lib/transform/include.test.ts (hunk 3 above)
Comment (JP-Dhabolt) [reply packages/config-loader/src/lib/transform/include.test.ts]: Made this a regular function.

File: packages/config-loader/src/lib/transform/types.ts [hunk 4]
```diff
@@ -23,13 +23,8 @@ export type ReadFileFunc = (path: string) => Promise<string>;
...
-      applied: true;
-      value: JsonValue | undefined;
-      newBaseDir?: string | undefined;
-    }
->;
+) => Promise<{
+  applied: boolean;
+  value?: JsonValue;
+  newBaseDir?: string;
+}>;
```
Comment (Rugvip) [inline packages/config-loader/src/lib/transform/types.ts:30]: Didn't spot this first time around, why is this change needed?

Comment (JP-Dhabolt) [reply packages/config-loader/src/lib/transform/types.ts:30]: It is simplifying the existing declaration into something that is functionally equivalent (no changes were made to any of the implementations of this type, and they all still compile).

Without this, there is some Typescript gymnastics that must be done for the `substitute` function in the `include.test.ts`.

With the previous declaration, `newBaseDir?: string | undefined;` compiled to `(string | undefined) | undefined`, and there were duplicate definitions of `applied: true` and `applied: false` (`applied: boolean`) that both had `value` and `newBaseDir` as `undefined` (either implicitly by leaving them out, or explicitly by declaring them as `string | undefined` or `JsonValue | undefined`).

If you would rather leave this type the same, I can revert and perform the necessary Typescript changes in `include.test.ts`, but I thought a refactor that simplified and didn't impact the functionality would be ok.  😄

Comment (Rugvip) [reply packages/config-loader/src/lib/transform/types.ts:30]: Would rather revert and do a bit of gymnastics in the test tbh. The old declaration guarantees makes it clearer that the extra fields aren't set if applied is `false`, and TypeScript might even be able to enforce that in some cases 😅

Comment (JP-Dhabolt) [reply packages/config-loader/src/lib/transform/types.ts]: Sure, no problem. Reverted.

File: packages/config-loader/src/lib/transform/include.test.ts [hunk 5]
```diff
@@ -14,25 +14,54 @@
...
+const substituteMe = '${MY_SUBSTITUTION}';
+const mySubstitution = 'fooSubstitution';
 
 const env = jest.fn(async (name: string) => {
   return ({
     SECRET: 'my-secret',
   } as { [name: string]: string })[name];
 });
 
+const substitute = async (
```
Comment (Rugvip) [inline packages/config-loader/src/lib/transform/include.test.ts:32]: Could prolly slim down the typing here using `const substitute: TransformFunc = ...`

Comment (JP-Dhabolt) [reply packages/config-loader/src/lib/transform/include.test.ts]: Done! 👍🏻   Sorry for the noise, not 100% with my Typescript just yet.

**Review by Rugvip — APPROVED** [review]
No worries, thanks!
